namesys/base.go

```diff
@@ -1,56 +1,111 @@
 package namesys
 
 import (
+	"context"
 	"strings"
 	"time"
 
-	context "context"
-
 	opts "github.com/ipfs/go-ipfs/namesys/opts"
+
 	path "gx/ipfs/QmdrpbDgeYH3VxkCciQCJY5LkDYdXtig6unDzQmMxFtWEw/go-path"
 )
 
+type onceResult struct {
+	value path.Path
+	ttl   time.Duration
+	err   error
+}
+
 type resolver interface {
-	// resolveOnce looks up a name once (without recursion).
-	resolveOnce(ctx context.Context, name string, options *opts.ResolveOpts) (value path.Path, ttl time.Duration, err error)
+	resolveOnceAsync(ctx context.Context, name string, options opts.ResolveOpts) <-chan onceResult
 }
 
 // resolve is a helper for implementing Resolver.ResolveN using resolveOnce.
-func resolve(ctx context.Context, r resolver, name string, options *opts.ResolveOpts, prefixes ...string) (path.Path, error) {
-	depth := options.Depth
-	for {
-		p, _, err := r.resolveOnce(ctx, name, options)
+func resolve(ctx context.Context, r resolver, name string, options opts.ResolveOpts, prefix string) (path.Path, error) {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	err := ErrResolveFailed
+	var p path.Path
+
+	resCh := resolveAsync(ctx, r, name, options, prefix)
+
+	for res := range resCh {
+		p, err = res.Path, res.Err
 		if err != nil {
-			return "", err
+			break
 		}
-		log.Debugf("resolved %s to %s", name, p.String())
+	}
 
-		if strings.HasPrefix(p.String(), "/ipfs/") {
-			// we've bottomed out with an IPFS path
-			return p, nil
-		}
+	return p, err
+}
 
-		if depth == 1 {
-			return p, ErrResolveRecursion
-		}
+func resolveAsync(ctx context.Context, r resolver, name string, options opts.ResolveOpts, prefix string) <-chan Result {
+	resCh := r.resolveOnceAsync(ctx, name, options)
+	depth := options.Depth
+	outCh := make(chan Result, 1)
+
+	go func() {
+		defer close(outCh)
+		var subCh <-chan Result
+		var cancelSub context.CancelFunc
 
-		matched := false
-		for _, prefix := range prefixes {
-			if strings.HasPrefix(p.String(), prefix) {
-				matched = true
-				if len(prefixes) == 1 {
-					name = strings.TrimPrefix(p.String(), prefix)
+		for {
+			select {
+			case res, ok := <-resCh:
+				if !ok {
+					resCh = nil
+					break
 				}
-				break
-			}
-		}
 
-		if !matched {
-			return p, nil
-		}
+				if res.err != nil {
+					outCh <- Result{Err: res.err}
+					return
+				}
+				log.Debugf("resolved %s to %s", name, res.value.String())
+				if strings.HasPrefix(res.value.String(), "/ipfs/") {
+					outCh <- Result{Path: res.value}
+					break
+				}
+
+				if depth == 1 {
+					outCh <- Result{Path: res.value, Err: ErrResolveRecursion}
+					break
+				}
+
+				subopts := options
+				if subopts.Depth > 1 {
+					subopts.Depth--
+				}
 
-		if depth > 1 {
-			depth--
+				var subCtx context.Context
+				if subCh != nil {
+					// Cancel previous recursive resolve since it won't be used anyways
+					cancelSub()
+				}
+				subCtx, cancelSub = context.WithCancel(ctx)
+				defer cancelSub()
```

Comment:
Defering in a loop is generally a bad idea (will allocate for each). Let's just make sure to call `cancelSub`. Maybe a `defer func() { if cancelSub != nil { cancelSub() }` at the top, or something like that.